internal/controller/workflow/controller.go

```diff
@@ -0,0 +1,592 @@
+		Name:      workflow.Spec.WorkflowDefinitionRef,
+		Namespace: workflow.Namespace,
+	}, workflowDef); err != nil {
+		logger.Error(err, "failed to get WorkflowDefinition")
+		return r.updateStatusAndRequeue(ctx, oldWorkflow, workflow)
+	}
+
+	_, workflowRunNamespace, err := r.getWorkflowRunMetadata(workflow, workflowDef)
+	if err != nil {
+		logger.Error(err, "failed to get workflow run name and namespace")
+		return r.updateStatusAndRequeue(ctx, oldWorkflow, workflow)
+	}
+
+	return r.ensureWorkflowResource(ctx, oldWorkflow, workflow, workflowDef, workflowRunNamespace, bpClient)
+}
+
+func (r *Reconciler) handleWorkloadCreation(
+	ctx context.Context,
+	oldWorkflow, workflow *openchoreodevv1alpha1.Workflow,
+) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	buildPlane, err := controller.GetBuildPlane(ctx, r.Client, workflow)
+	if err != nil {
+		logger.Error(err, "failed to get build plane for workload creation")
+		return r.updateStatusAndRequeue(ctx, oldWorkflow, workflow)
+	}
+
+	bpClient, err := r.getBuildPlaneClient(buildPlane)
+	if err != nil {
+		logger.Error(err, "failed to get build plane client for workload creation")
+		return r.updateStatusAndRequeue(ctx, oldWorkflow, workflow)
+	}
+
+	shouldRequeue, err := r.createWorkloadFromWorkflowRun(ctx, workflow, bpClient)
+	if err != nil {
+		logger.Error(err, "failed to create workload CR")
+		if shouldRequeue {
+			return r.updateStatusAndRequeue(ctx, oldWorkflow, workflow)
+		}
+		setWorkloadUpdateFailedCondition(workflow)
+		return r.updateStatusAndReturn(ctx, oldWorkflow, workflow)
+	}
+
+	setWorkloadUpdatedCondition(workflow)
+	return r.updateStatusAndReturn(ctx, oldWorkflow, workflow)
+}
+
+func (r *Reconciler) ensureWorkflowResource(
+	ctx context.Context,
+	oldWorkflow, workflow *openchoreodevv1alpha1.Workflow,
+	workflowDef *openchoreodevv1alpha1.WorkflowDefinition,
+	workflowRunNamespace string,
+	bpClient client.Client,
+) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	component := &openchoreodevv1alpha1.Component{}
+	if err := r.Get(ctx, types.NamespacedName{
+		Name:      workflow.Spec.Owner.ComponentName,
+		Namespace: workflow.Namespace,
+	}, component); err != nil {
+		logger.Error(err, "failed to get Component")
+		return r.updateStatusAndRequeue(ctx, oldWorkflow, workflow)
+	}
+
+	componentTypeDefName, err := parseComponentTypeDefinitionName(component.Spec.ComponentType)
+	if err != nil {
+		logger.Error(err, "failed to parse ComponentTypeDefinition name")
+		return r.updateStatusAndRequeue(ctx, oldWorkflow, workflow)
+	}
+
+	componentTypeDef := &openchoreodevv1alpha1.ComponentTypeDefinition{}
+	if err := r.Get(ctx, types.NamespacedName{
+		Name:      componentTypeDefName,
+		Namespace: workflow.Namespace,
+	}, componentTypeDef); err != nil {
+		logger.Error(err, "failed to get ComponentTypeDefinition")
+		return r.updateStatusAndRequeue(ctx, oldWorkflow, workflow)
+	}
+
+	pipeline := workflowpipeline.NewPipeline()
```

Comment:
Should be initialized once per Reconciler instance. The template engine maintains a cache of compiled expressions, so it needs to be reused across reconciliations.

https://github.com/openchoreo/openchoreo/blob/8ef8ae2de1ba91e0cb27aa1a6f03edc906f94a93/internal/controller/componentdeployment/controller.go#L32-L39